mono/metadata/android-ifaddrs.c

```diff
@@ -0,0 +1,1158 @@
+				goto error;
+			}
+		}
+
+		attribute = RTA_NEXT (attribute, length);
+	}
+
+	/* glibc stores the associated interface name in the address if IFA_LABEL never occured */
+	if (!ifa->ifa_name) {
+		gchar *name = get_interface_name_by_index (net_address->ifa_index, ifaddrs_head);
+		g_debug ("   address has no name/label, getting one from interface\n");
+		ifa->ifa_name = name ? strdup (name) : NULL;
+	}
+	g_debug ("   address label: %s\n", ifa->ifa_name);
+
+	if (calculate_address_netmask (ifa, net_address) < 0) {
+		goto error;
+	}
+
+	return ifa;
+
+  error:
+	{
+		/* errno may be modified by free, or any other call inside the free_single_xamarin_ifaddrs
+		 * function. We don't care about errors in there since it is more important to know how we
+		 * failed to obtain the link address and not that we went OOM. Save and restore the value
+		 * after the resources are freed.
+		 */
+		gint errno_save = errno;
+		free_single_xamarin_ifaddrs (&ifa);
+		errno = errno_save;
+		return NULL;
+	}
+
+}
+
+static struct _monodroid_ifaddrs *
+get_link_info (const struct nlmsghdr *message)
+{
+	ssize_t length;
+	struct rtattr *attribute;
+	struct ifinfomsg *net_interface;
+	struct _monodroid_ifaddrs *ifa = NULL;
+	struct sockaddr_ll_extended *sa = NULL;
+
+	g_assert (message);
+	net_interface = NLMSG_DATA (message);
+	length = message->nlmsg_len - NLMSG_LENGTH (sizeof (*net_interface));
+	if (length <= 0) {
+		goto error;
+	}
+
+	ifa = calloc (1, sizeof (*ifa));
+	if (!ifa) {
+		goto error;
+	}
+
+	ifa->ifa_flags = net_interface->ifi_flags;
+	attribute = IFLA_RTA (net_interface);
+	while (RTA_OK (attribute, length)) {
+		switch (attribute->rta_type) {
+			case IFLA_IFNAME:
+				ifa->ifa_name = strdup (RTA_DATA (attribute));
+				if (!ifa->ifa_name) {
+					goto error;
+				}
+				g_debug ("   interface name (payload length: %d; string length: %d)\n", RTA_PAYLOAD (attribute), strlen (ifa->ifa_name));
+				g_debug ("     %s\n", ifa->ifa_name);
+				break;
+
+			case IFLA_BROADCAST:
+				g_debug ("   interface broadcast (%d bytes)\n", RTA_PAYLOAD (attribute));
+				if (fill_ll_address (&sa, net_interface, RTA_DATA (attribute), RTA_PAYLOAD (attribute)) < 0) {
+					goto error;
+				}
+				ifa->_monodroid_ifa_broadaddr = (struct sockaddr*)sa;
+				break;
+
+			case IFLA_ADDRESS:
+				g_debug ("   interface address (%d bytes)\n", RTA_PAYLOAD (attribute));
+				if (fill_ll_address (&sa, net_interface, RTA_DATA (attribute), RTA_PAYLOAD (attribute)) < 0) {
+					goto error;
+				}
+				ifa->ifa_addr = (struct sockaddr*)sa;
+				break;
+
+			default:
+				g_debug ("     rta_type: ");
+#if DEBUG
+				print_ifla_name (attribute->rta_type);
+#endif
+				break;
+		}
+
+		attribute = RTA_NEXT (attribute, length);
+	}
+	g_debug ("link flags: 0x%X", ifa->ifa_flags);
+	return ifa;
+
+  error:
+	if (sa)
+		free (sa);
+	free_single_xamarin_ifaddrs (&ifa);
+
+	return NULL;
+}
+
+#if DEBUG
+
+static void
+print_ifla_name (gint id)
+{
+	static struct {
```

Comment:
Probably should be const. I see it is under if DEBUG.

Reply:
When/if we move to C++, yes, in C it doesn't matter.

Reply:
It matters the same in C. You want read only data on read only pages and writable data on  writable pages.

So when you write to actually writable data, the page fault only brings in other writable data, instead of wastefully forking the page for readonly data.

This minimizes copy-on-write pages faults. Assuming mobile works like normal.

On systems without demand paging, it doesn't matter nearly as much, they read it all in up front. It can still help with efficient cache coherency on multi-proc systems but these are usually smaller effects (bouncing written cache line ownership across cores vs. shared ownership of readonly).